eodag/plugins/search/cop_marine.py

```diff
@@ -348,16 +372,54 @@ def query(
 
                 for obj in s3_objects["Contents"]:
                     item_key = obj["Key"]
+                    item_id = item_key.split("/")[-1].split(".")[0]
                     # filter according to date(s) in item id
-                    item_dates = re.findall(r"\d{8}", item_key)
+                    item_dates = re.findall(r"(\d{4})(0[1-9]|1[0-2])([0-3]\d)", item_id)
                     if not item_dates:
-                        item_dates = re.findall(r"\d{6}", item_key)
-                    item_start = _get_date_from_yyyymmdd(item_dates[0], item_key)
-                    if not item_start:  # identified pattern was not a valid datetime
+                        item_dates = re.findall(r"_(\d{4})(0[1-9]|1[0-2])", item_id)
+                    item_dates = [
+                        "".join(row) for row in item_dates
+                    ]  # join tuples returned by findall
+                    item_start = None
+                    item_end = None
+                    use_dataset_dates = False
+                    if item_dates:
+                        item_start = _get_date_from_yyyymmdd(item_dates[0], item_key)
+                        if len(item_dates) > 2:  # start, end and created_at timestamps
+                            item_end = _get_date_from_yyyymmdd(item_dates[1], item_key)
+                    if not item_start:
+                        # no valid datetime given in id
+                        use_dataset_dates = True
+                        dates = _get_dates_from_dataset_data(dataset_item)
+                        if dates:
+                            item_start = dates["start"].replace("Z", "+0000")
+                            item_end = dates["end"].replace("Z", "+0000")
```

Comment:
Incompatible types in assignment (expression has type "str", variable has type "datetime | None")mypy[assignment](https://mypy.readthedocs.io/en/stable/_refs.html#code-assignment)